### ISSUE TITLE
Subband vs Duration

### ISSUE THREAD

**Issue by zaptrem** [bfs18/rfwave#3]
Hi, can you explain the difference between the subband and duration experiments and share which you've found to perform better? Also, what have you discovered about the use of classifier free guidance?

**Comment by bfs18** [comment]
> Hi, can you explain the difference between the subband and duration experiments and share which you've found to perform better? Also, what have you discovered about the use of classifier free guidance?

Thank you for your interest. The term 'duration' pertains to text-to-speech applications and is unrelated to the reconstruction of audio waveforms from Mel-spectrograms or EnCodec tokens. Regarding the number of subbands, our preliminary experiments indicated that using 16 subbands yields better results than 4 subbands, and 4 subbands outperform full-band processing. Employing more subbands increases the amount of computation, while maintaining a similar count of parameters and utilizing the same dataset. This approach could enhance performance according to the empirical scaling law, although it's worth noting that the scaling law has been primarily summarized from experiments with Transformer models. I've attached a figure to illustrate classifier-free guidance and STFT loss. While CFG enhances objective metrics for vocoder experiments, it does not lead to a corresponding increase in listening test scores; however, when reconstructing waveforms from EnCodec tokens, CFG substantially improves the listening experience. This may be because the abundant and deterministic information in the Mel-spectrogram renders CFG unnecessary.
![09f21e15-74c2-4f60-bde3-442e6aaf072f](https://github.com/bfs18/rfwave/assets/39617376/be2ee672-be43-4fdd-8003-cf6d21867d62)

**Comment by zaptrem** [comment]
Thanks! A few more questions:

1. You mention 4 vs 16 subbands but I think the paper and code use 8 subbands. Is there a reason you're discussing 4 vs 16 here? 
2. I see you're using STFT loss in the above which is off by default in the code. Did you find phase loss and overlap loss to produce worse results? Is energy-balanced loss on by default?
3. Have you tried applying your multi-band approach to Vocos (i.e., no diffusion)?
4. Have you noticed that models with CFG turned on are significantly louder than those with it off?
5. I noticed some buzzing when the audio is quiet/silent even with time balanced loss turned on. Does this go away as training continues?
6. I'm testing out your models but have been getting unusual loss curves. Did you see these as well? 
![image](https://github.com/bfs18/rfwave/assets/1612230/dd0ac857-ed2e-4e54-ad30-2719d37384d0)

**Comment by bfs18** [comment]
Hi @zaptrem ，

1. Apologies for the confusion earlier. In our initial experiments, we actually tested performance with 4 vs 8 subbands, not 16 subbands.
2. The use of STFT loss has demonstrated an advantage in mitigating water-like noise when background noise is present in our experiments. We did not observe any significant improvements by incorporating phase loss. Moreover, the incorporation of overlap loss does not compromise performance; rather, it ensures coherence among the individually modeled subbands. We have determined that a weight coefficient of 0.01 is sufficient for both STFT loss and overlap loss in our setup. Energy balanced loss is discussed in Q5.
The following is an example for STFT loss. There are some vertical patterns in the spectrogram of waveforms generated by a model without STFT loss.
[background.zip](https://github.com/user-attachments/files/15519487/background.zip)

Groundth
<img width="414" alt="gt" src="https://github.com/bfs18/rfwave/assets/39617376/f9af51e5-ec68-4790-b65c-faba3f62dd62">

With STFT loss
<img width="414" alt="stft" src="https://github.com/bfs18/rfwave/assets/39617376/dc831002-0a17-41f3-a181-d419cbdb8212">

Without STFT loss
<img width="414" alt="wo stft" src="https://github.com/bfs18/rfwave/assets/39617376/8d0afd20-1aa0-476d-ba14-20bc49ee08ef">

3. We have not yet applied the multi-band approach to the Vocos system. However, I believe it could be beneficial for Vocos as well, given that the multi-band approach generally leads to increased computational load.
4. I noticed that. Training input audio is normalized to a range between -1 and -6 dB. Therefore, normalizing the audio to this volume level during testing should yield more consistent results.
5. Energy-balanced loss (time balanced loss in code) is designed for this issue. It will disappear as training continues.
6. This is my loss curve on LJSpeech dataset (22.05 kHz).
![ae2c1f4e-1557-47c0-96e5-f9816f38c63a](https://github.com/bfs18/rfwave/assets/39617376/3373e7da-d609-4025-8aa0-194e08281eb2)

On Opencpop dataset (44.1 kHz).
![02624bbb-1dc7-41a9-9a82-c8f98ded9496](https://github.com/bfs18/rfwave/assets/39617376/4deaaa42-8616-48af-b217-7a815b68fed6)

**Comment by zaptrem** [comment]
Thanks! It looks like you probably have a similar staircase effect between 0 and 30k but I'm not certain since it's zoomed out. Also, I noticed your PQMF filter is hard-coded to 8 bands, 124 taps, and cutoff 0.071. If using 16 bands would it be better to set these to (following the trend you set going from 4 to 8) 16 bands, 248 taps, and cutoff 0.0355? Also, similar to CFG scale have you noticed any other changes that disproportionately help with generating waveforms for Encodec tokens?

**Comment by bfs18** [comment]
Hi @zaptrem ,
The PQMF is only used for waveform equalization, and 4 subbands are utilized. These subbands are equalized and then merged into an equalized waveform. The model splits the complex spectrogram into 8 subbands by selecting the appropriate dimensions and it has nothing to do with PQMF. 
I haven't experimented with splitting the complex spectrogram into 16 subbands, and I haven't observed any other factors that disproportionately enhance the generation of waveforms for Encodec tokens.

**Comment by zaptrem** [comment]
Thanks! For clarification, the original paper used 4 vs 8 (model, not PQMF) subbands, but you have since moved to 16 and that is how you got the results in [this](https://private-user-images.githubusercontent.com/39617376/334339612-be2ee672-be43-4fdd-8003-cf6d21867d62.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1Mzc1NjUsIm5iZiI6MTcxNzUzNzI2NSwicGF0aCI6Ii8zOTYxNzM3Ni8zMzQzMzk2MTItYmUyZWU2NzItYmU0My00ZmRkLTgwMDMtY2Y2ZDIxODY3ZDYyLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwNFQyMTQxMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMjlkZDY3NjViNzJjYTAxMWNmMWY2OWJmMDNjOGM0OGRhMjc2YjNmOWU1YTZmZDlkY2RhYTNjOTFmMmM5Njc5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.p4w-HrTw9zxB2ma2t8cOPt3g0R-uYZAkoMJrWoSO9yQ) screenshot? Also, when you switched between 4/8/16 model subbands did you need to adjust the `left_overlap` and `right_overlap` parameters (which are both set to 8 by default) in `RectifiedFlow`?

**Comment by bfs18** [comment]
> Thanks! For clarification, the original paper used 4 vs 8 (model, not PQMF) subbands, but you have since moved to 16 and that is how you got the results in [this](https://private-user-images.githubusercontent.com/39617376/334339612-be2ee672-be43-4fdd-8003-cf6d21867d62.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1Mzc1NjUsIm5iZiI6MTcxNzUzNzI2NSwicGF0aCI6Ii8zOTYxNzM3Ni8zMzQzMzk2MTItYmUyZWU2NzItYmU0My00ZmRkLTgwMDMtY2Y2ZDIxODY3ZDYyLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwNFQyMTQxMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMjlkZDY3NjViNzJjYTAxMWNmMWY2OWJmMDNjOGM0OGRhMjc2YjNmOWU1YTZmZDlkY2RhYTNjOTFmMmM5Njc5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.p4w-HrTw9zxB2ma2t8cOPt3g0R-uYZAkoMJrWoSO9yQ) screenshot? Also, when you switched between 4/8/16 model subbands did you need to adjust the `left_overlap` and `right_overlap` parameters (which are both set to 8 by default) in `RectifiedFlow`?

Hi, 
1. The image link didn't work out. I can't see it. Could you resend it?
2.  The 8-dimensional overlap is fine for varying subbands. Adjustments to left_overlap and right_overlap aren't typically needed when changing subbands.

**Comment by zaptrem** [comment]
> > Thanks! For clarification, the original paper used 4 vs 8 (model, not PQMF) subbands, but you have since moved to 16 and that is how you got the results in [this](https://private-user-images.githubusercontent.com/39617376/334339612-be2ee672-be43-4fdd-8003-cf6d21867d62.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1Mzc1NjUsIm5iZiI6MTcxNzUzNzI2NSwicGF0aCI6Ii8zOTYxNzM3Ni8zMzQzMzk2MTItYmUyZWU2NzItYmU0My00ZmRkLTgwMDMtY2Y2ZDIxODY3ZDYyLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwNFQyMTQxMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMjlkZDY3NjViNzJjYTAxMWNmMWY2OWJmMDNjOGM0OGRhMjc2YjNmOWU1YTZmZDlkY2RhYTNjOTFmMmM5Njc5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.p4w-HrTw9zxB2ma2t8cOPt3g0R-uYZAkoMJrWoSO9yQ) screenshot? Also, when you switched between 4/8/16 model subbands did you need to adjust the `left_overlap` and `right_overlap` parameters (which are both set to 8 by default) in `RectifiedFlow`?
> 
> Hi,
> 
> 1. The image link didn't work out. I can't see it. Could you resend it?
> 2. The 8-dimensional overlap is fine for varying subbands. Adjustments to left_overlap and right_overlap aren't typically needed when changing subbands.


2. Thanks! 
1. This one: 
![image](https://github.com/bfs18/rfwave/assets/1612230/44e344dc-67ab-4f14-846f-55d78e3b58aa)

Were these with 16 bands? If not, did 16 bands improve it beyond these results?

**Comment by bfs18** [comment]
> > > Thanks! For clarification, the original paper used 4 vs 8 (model, not PQMF) subbands, but you have since moved to 16 and that is how you got the results in [this](https://private-user-images.githubusercontent.com/39617376/334339612-be2ee672-be43-4fdd-8003-cf6d21867d62.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1Mzc1NjUsIm5iZiI6MTcxNzUzNzI2NSwicGF0aCI6Ii8zOTYxNzM3Ni8zMzQzMzk2MTItYmUyZWU2NzItYmU0My00ZmRkLTgwMDMtY2Y2ZDIxODY3ZDYyLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwNFQyMTQxMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMjlkZDY3NjViNzJjYTAxMWNmMWY2OWJmMDNjOGM0OGRhMjc2YjNmOWU1YTZmZDlkY2RhYTNjOTFmMmM5Njc5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.p4w-HrTw9zxB2ma2t8cOPt3g0R-uYZAkoMJrWoSO9yQ) screenshot? Also, when you switched between 4/8/16 model subbands did you need to adjust the `left_overlap` and `right_overlap` parameters (which are both set to 8 by default) in `RectifiedFlow`?
> > 
> > 
> > Hi,
> > 
> > 1. The image link didn't work out. I can't see it. Could you resend it?
> > 2. The 8-dimensional overlap is fine for varying subbands. Adjustments to left_overlap and right_overlap aren't typically needed when changing subbands.
> 
> 2. Thanks!
> 3. This one:
>    ![image](https://private-user-images.githubusercontent.com/1612230/336647334-44e344dc-67ab-4f14-846f-55d78e3b58aa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NTMzNzAsIm5iZiI6MTcxNzU1MzA3MCwicGF0aCI6Ii8xNjEyMjMwLzMzNjY0NzMzNC00NGUzNDRkYy02N2FiLTRmMTQtODQ2Zi01NWQ3OGUzYjU4YWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDVUMDIwNDMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTQ2ZTM4MjY2MWQxMDhjZDQ1YjdjYjhjZWU3OWU4YTgwYWZlYmY0NTllZDQ4OTgxZmY3MjhkNzY0ZTAyMDRlNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.uA4cCq81UWjiv9loaQH7M0gKwzjtn1HMu8OJ14VRhLQ)
> 
> Were these with 16 bands? If not, did 16 bands improve it beyond these results?

In this table, RFWave utilizes 8 subbands, a detail noted in the paper's text but omitted from the table's title.

**Comment by zaptrem** [comment]
![image](https://github.com/bfs18/rfwave/assets/1612230/bb8e4e20-b012-46ee-9f89-68929d38d591)


Doesn't this table (and yours above) show overlap loss actually making the results _worse_ (PESQ, V/UV, Periodicity) or no better (ViSQOL) compared to no overlap loss?

**Comment by bfs18** [comment]
> ![image](https://private-user-images.githubusercontent.com/1612230/343988501-bb8e4e20-b012-46ee-9f89-68929d38d591.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1NjYyNDksIm5iZiI6MTcxOTU2NTk0OSwicGF0aCI6Ii8xNjEyMjMwLzM0Mzk4ODUwMS1iYjhlNGUyMC1iMDEyLTQ2ZWUtOWY4OS02ODkyOWQzOGQ1OTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjhUMDkxMjI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmU1ZjRkOWJhOGM5ZWM4OWZjYTIxZGYzZjYyNTRlNzBhMjNkZTNjZjAzNmI3NjhiYjQ0ZjM5NWY3ZGEyMWExZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.qi2G8NrAQW2JFmgGuiPxQQVBxuPTBK-ahOSRTbZBsG8)
> 
> Doesn't this table (and yours above) show overlap loss actually making the results _worse_ (PESQ, V/UV, Periodicity) or no better (ViSQOL) compared to no overlap loss?

In initial trials, I noticed the presence of horizontal striations across subbands within the spectrograms occasionally. Implementing overlap loss as a countermeasure effectively mitigated this issue, leading to its adoption as the default setting.  I did not expect an improvement in objective metrics with overlap loss. My primary goal was to ensure robust performance across a diverse range of configurations.

**Comment by zaptrem** [comment]
Thanks. Your model uses a lot of bands/compute on frequencies the human ear doesn't really care about, so I've been trying to fix that. Here's what 16 bands looks like with your current approach:

![standard_spectrogram](https://github.com/bfs18/rfwave/assets/1612230/4327506b-0f9b-4fe4-b0e2-8b234a085fce)


I thought of using an stft with a higher n_fft (4096) for the lower frequencies (so more bands/compute is spent on parts we care about) and a lower n_fft on the higher frequencies (1024 but still satisfying COLA since hop size is 512).

![stft_comparison_50_iterations](https://github.com/bfs18/rfwave/assets/1612230/22dd1840-07b5-48c3-a7e1-da981cb879ea)


The problem I ran into is my multi-resolution STFT implementation is not as perfectly invertible as a normal spectrogram (though it's close), and I'm not quite sure why. When you run it back and forth hundreds of times some ringing/artifacts will appear on the border between the two resolutions. When you train on these (with overlap turned off since it doesn't make sense across resolution boundaries and wave=true, haven't tried false) the model seems to have a much harder time wrapping its head around anything and after one night still hasn't figured out phase.

Have you tried anything like this?

My code in case you're interested: https://gist.github.com/zaptrem/94d10c5d76d2f601841e9f8e8bf4859a




Also, I'm not quite sure I understand the motivation for doing this `pred = self.stft(self.istft(pred))` all the time (e.g., in compute_stft_loss and when taking inference steps) when wave is true. Why do you do it?

And why did you stop using the trick [Vocos uses ](https://github.com/gemelo-ai/vocos/blob/c859e3b7b534f3776a357983029d34170ddd6fc3/vocos/heads.py#L55)for better phase outputs (since phase is periodic)?

Also, (sorry for so many questions haha) with wave=false did you find the feature_loss to improve things or make them worse?

**Comment by bfs18** [comment]
Hi @zaptrem 
1. I've tested your code and noticed a horizontal line between the two subbands. Additionally, there seems to be an error in the code regarding the subband dimensions; they should be 1024 and 257 instead of 1024 and 256. However, even after correcting this, there are still artifacts present. I have not yet determined a solution to this issue. One potential approach to mitigate the error accumulation could be to set wave=False and conduct the modeling directly in the frequency domain. This would necessitate only a single inverse STFT operation after sampling, which may reduce the severity of artifacts.
![20240701-144528](https://github.com/bfs18/rfwave/assets/39617376/ca4cb05a-dfd9-40d8-9979-db2a1df11eb3)
3. stft(istft(pred)) follows the stft and istft operation in Figure 1 in the paper.

**Comment by zaptrem** [comment]
> 1. Additionally, there seems to be an error in the code regarding the subband dimensions; they should be 1024 and 257 instead of 1024 and 256

Thanks! Can you tell me more about where this is happening and how you fixed it? One of my suspicions for why this was performing significantly worse was the shape/location of the real/imaginary channels was shifting between bands, but I could never determine whether that was actually happening. Also noting here that I tried some CQT/wavelet based transforms and got similarly mediocre results, I think because they're much more periodic than real/imag STFT. 

Also, I've noticed the waterfall effect seems to continue with even with STFT loss enabled in high-noise regions of the spectrogram and it becomes more prominent as you get further in the training run. Could it be caused by errors being amplified between bands via the stft/istft operation from each step?

**Comment by zaptrem** [comment]
![image](https://github.com/bfs18/rfwave/assets/1612230/e5a692a2-8d00-47ca-8c57-0f1c3e3c8290)

Also, I noticed the waterfall effect is still quite prevalent when there's noise but not complete silence even at the end of training with STFT loss enabled.

**Comment by bfs18** [comment]
> ![image](https://private-user-images.githubusercontent.com/1612230/347281945-e5a692a2-8d00-47ca-8c57-0f1c3e3c8290.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1OTk2NjMsIm5iZiI6MTcyMDU5OTM2MywicGF0aCI6Ii8xNjEyMjMwLzM0NzI4MTk0NS1lNWE2OTJhMi04ZDAwLTQ3Y2EtOGM1Ny0wZjFjM2UzYzgyOTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTBUMDgxNjAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmUzODdhNzIyNWU2NTExYWNmNGUxZDkzY2NiYmE3ZjQwOGMzNDU1NDU2NGM3N2QwNGU5YjFhYmYzMmM4YTdkMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7FTKNdCmgfc13DYy7Q4BdA9Yks9op5FAlhPJajgT3VQ)
> 
> Also, I noticed the waterfall effect is still quite prevalent when there's noise but not complete silence even at the end of training with STFT loss enabled.
Hi @zaptrem 
I also don't know how to fix the mismatch when shifting between bands of different FFT size.
Regarding the waterfall noise, I believe it occurs because the model attempts to reconstruct a phase even for background noise, where no meaningful phase is present. Increasing the STFT loss weight might resolve this issue by making the model place slightly more emphasis on the magnitude information.

**Comment by zaptrem** [comment]
> > ![image](https://private-user-images.githubusercontent.com/1612230/347281945-e5a692a2-8d00-47ca-8c57-0f1c3e3c8290.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1OTk2NjMsIm5iZiI6MTcyMDU5OTM2MywicGF0aCI6Ii8xNjEyMjMwLzM0NzI4MTk0NS1lNWE2OTJhMi04ZDAwLTQ3Y2EtOGM1Ny0wZjFjM2UzYzgyOTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTBUMDgxNjAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmUzODdhNzIyNWU2NTExYWNmNGUxZDkzY2NiYmE3ZjQwOGMzNDU1NDU2NGM3N2QwNGU5YjFhYmYzMmM4YTdkMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7FTKNdCmgfc13DYy7Q4BdA9Yks9op5FAlhPJajgT3VQ)
> 
> > 
> 
> > Also, I noticed the waterfall effect is still quite prevalent when there's noise but not complete silence even at the end of training with STFT loss enabled.
> 
> Hi @zaptrem 
> 
> I also don't know how to fix the mismatch when shifting between bands of different FFT size.
> 
> Regarding the waterfall noise, I believe it occurs because the model attempts to reconstruct a phase even for background noise, where no meaningful phase is present. Increasing the STFT loss weight might resolve this issue by making the model place slightly more emphasis on the magnitude information.
> 
> 


I figured out you can use a PQMF to get clean cuts that you can apply different STFT settings to, but I haven't tried a model with it yet because idk how the aliasing cancellation will work when neither top-level-band is aware of the other.

Something I noticed (though haven't rigorously confirmed) is that the waterfall effect is actually *less* prominent earlier in training. Also notice how the lines span between bands. Could small errors be getting amplified by the overlap between bands or possibly some sort of spectral bleeding during sampling? Did you see the waterfall effect in your experiments from before you started using STFT(ISTFT()) each step during sampling?


Edit: The waterfalling goes away entirely when I disable the stft(istft()) at inference time. However, otherwise the quality becomes worse:
![image](https://github.com/bfs18/rfwave/assets/1612230/388736a0-ee64-43f1-85c5-bf2ae46578ae)


Top: istft/stft turned off. Middle: Turned on. Bottom: Ground truth.

**Comment by zaptrem** [comment]
@accum-dai @Beronx86 Have you guys noticed this? It is possible solving this issue could help RFWave beat BigVGAN.

**Comment by bfs18** [comment]
> Something I noticed (though haven't rigorously confirmed) is that the waterfall effect is actually _less_ prominent earlier in training. Also notice how the lines s

Hi @zaptrem , 
I apologize for not noticing that you updated your post. "stft(istft(x))" is added when the model supports conditioning higher bands on lower bands and is applied on a subband's complex spectrum. This is because the inverse short-time Fourier transform (istft) of a crafted spectrum may not yield a real signal, which causes error accumulation when taking the inverse as real and applying many sampling steps. When predicting all subbands in parallel, "stft(istft(x))" actually operates on the full band spectrum, making it potentially unnecessary. However, I have not yet attempted to remove it. I appreciate that you tried it, and I will conduct further tests to explore this.

**Comment by bfs18** [comment]
Hi @zaptrem Is your dataset publicly available? I would like to test it.

**Comment by zaptrem** [comment]
> Hi @zaptrem Is your dataset publicly available? I would like to test it.

It's a personal dataset so unfortunately can't share it. Closest approximation would be something like https://github.com/mdeff/fma plus https://github.com/MTG/mtg-jamendo-dataset.

**Comment by bfs18** [comment]
> It's a personal dataset so unfortunately can't share it. Closest approximation would be something like https://github.com/mdeff/fma plus https://github.com/MTG/mtg-jamendo-dataset.

@zaptrem Is it possible to share a few samples for testing, especially the one shown in the figure above?

**Comment by zaptrem** [comment]
I wish I could, but I'm not certain the license grants me permission to post it publicly. Here's a sample I made myself that should repro the issue

https://github.com/user-attachments/assets/d5d314c4-d7c0-4de2-9680-2e4a75978144

**Comment by bfs18** [comment]
> I wish I could, but I'm not certain the license grants me permission to post it publicly. Here's a sample I made myself that should repro the issue
> 
>  audiotest.mp4

Using [this checkpoint](https://drive.google.com/file/d/1yM0BWFrXvuwb2SZvyPOBr6Wgto0sjdqh/view), I haven't reproduce your issue. Could you test it on your internal test case and update the result? @zaptrem 
gt
<img width="487" alt="gt" src="https://github.com/user-attachments/assets/50b45363-9153-43a2-b5be-922a179cbdc5">

syn
<img width="487" alt="syn" src="https://github.com/user-attachments/assets/8e872dc2-19c6-443e-b158-de987efe34f4">

**Comment by Ted-Picovoice** [comment]
> Thanks! A few more questions:
> 
> 1. You mention 4 vs 16 subbands but I think the paper and code use 8 subbands. Is there a reason you're discussing 4 vs 16 here?
> 2. I see you're using STFT loss in the above which is off by default in the code. Did you find phase loss and overlap loss to produce worse results? Is energy-balanced loss on by default?
> 3. Have you tried applying your multi-band approach to Vocos (i.e., no diffusion)?
> 4. Have you noticed that models with CFG turned on are significantly louder than those with it off?
> 5. I noticed some buzzing when the audio is quiet/silent even with time balanced loss turned on. Does this go away as training continues?
> 6. I'm testing out your models but have been getting unusual loss curves. Did you see these as well?
>    ![image](https://github.com/bfs18/rfwave/assets/1612230/dd0ac857-ed2e-4e54-ad30-2719d37384d0)

For 6 I think it's just optimizer warm up.
